packet_buffer.go

```diff
@@ -121,19 +123,23 @@ func (pb *packetBuffer) next() (p *Packet, err error) {
 		pb.packetReadBuffer = make([]byte, pb.packetSize)
 	}
 
-	if _, err = io.ReadFull(pb.r, pb.packetReadBuffer); err != nil {
-		if err == io.EOF || err == io.ErrUnexpectedEOF {
-			err = ErrNoMorePackets
-		} else {
-			err = fmt.Errorf("astits: reading %d bytes failed: %w", pb.packetSize, err)
+	// We must return some results to user even if current packet has been filtered
+	for p == nil {
+		if _, err = io.ReadFull(pb.r, pb.packetReadBuffer); err != nil {
+			if err == io.EOF || err == io.ErrUnexpectedEOF {
+				err = ErrNoMorePackets
+			} else {
+				err = fmt.Errorf("astits: reading %d bytes failed: %w", pb.packetSize, err)
+			}
+			return
 		}
-		return
-	}
 
-	// Parse packet
-	if p, err = parsePacket(astikit.NewBytesIterator(pb.packetReadBuffer)); err != nil {
-		err = fmt.Errorf("astits: building packet failed: %w", err)
-		return
+		// Parse packet
+		if p, err = parsePacket(astikit.NewBytesIterator(pb.packetReadBuffer), pb.filter); err != nil {
```

Comment:
You'll need to add `&& !errors.Is(errSkippedPacket)` after `err != nil`

Reply:
Fixed with additional `if` for the sake of line shortening